### PR TITLE
PMM-6395 VMagent integration

### PR DESCRIPTION
**Pull request by f41gh7**
[PMM-6395](https://jira.percona.com/browse/PMM-6395)

- [x] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1185

- [x] Links to other linked pull requests (optional).
  - https://github.com/percona/pmm/pull/603
  - https://github.com/percona/pmm-admin/pull/122
  - https://github.com/percona/pmm-agent/pull/201
  - https://github.com/percona/pmm-managed/pull/501
  - https://github.com/percona/grafana-dashboards/pull/1034
  - https://github.com/Percona-Lab/pmm-api-tests/pull/140
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/converters.go [hunk 1]
```diff
@@ -202,12 +202,13 @@ func ToAPIAgent(q *reform.Querier, agent *models.Agent) (inventorypb.Agent, erro
...
-			Status:       inventorypb.AgentStatus(inventorypb.AgentStatus_value[agent.Status]),
-			ListenPort:   uint32(pointer.GetUint16(agent.ListenPort)),
-			CustomLabels: labels,
+			AgentId:             agent.AgentID,
+			PmmAgentId:          pointer.GetString(agent.PMMAgentID),
+			Disabled:            agent.Disabled,
+			Status:              inventorypb.AgentStatus(inventorypb.AgentStatus_value[agent.Status]),
+			ListenPort:          uint32(pointer.GetUint16(agent.ListenPort)),
+			CustomLabels:        labels,
+			PushMetricsDisabled: !agent.PushMetrics,
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/converters.go]: changed to push_metrics_enabled

File: services/inventory/agents.go [hunk 2]
```diff
@@ -80,12 +78,22 @@ func (as *AgentsService) changeAgent(agentID string, common *inventorypb.ChangeC
 		if got > 1 {
 			return status.Errorf(codes.InvalidArgument, "expected at most one param: enable or disable")
 		}
+		// cannot use both at the same time
+		if common.EnablePushMetrics && common.DisablePushMetrics {
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/inventory/agents.go]: fixed

File: models/agent_helpers.go [hunk 3]
```diff
@@ -451,6 +457,9 @@ func ChangeAgent(q *reform.Querier, agentID string, params *ChangeCommonAgentPar
 			row.Disabled = false
 		}
 	}
+	if params.DisablePushMetrics != nil {
+	row.PushMetrics = !(*params.DisablePushMetrics)
```
Comment (BupycHuk) [inline models/agent_helpers.go:461]: wrong indentation

File: services/victoriametrics/victoriametrics.go [hunk 4]
```diff
@@ -360,3 +349,22 @@ func scrapeConfigForVMAlert(interval time.Duration) *config.ScrapeConfig {
 		},
 	}
 }
+
+// BuildScrapeConfigForAgent - builds scrape configuration for given pmmAgent
+func (svc *Service) BuildScrapeConfigForAgent(pmmAgentID string) ([]byte, error) {
```
Comment (BupycHuk) [inline services/victoriametrics/victoriametrics.go:354]:
```suggestion
func (svc *Service) BuildScrapeConfigForVMAgent(pmmAgentID string) ([]byte, error) {
```

**Comment by AlekSi** [comment]
@f41gh7 "Drop Prometheus" PR is merged, please update

File: services/victoriametrics/victoriametrics.go (hunk 4 above)
Comment (f41gh7) [reply services/victoriametrics/victoriametrics.go]: updated

File: models/agent_helpers.go (hunk 3 above)
Comment (f41gh7) [reply models/agent_helpers.go]: yep, after suggestion, fixed.

**Review by BupycHuk — APPROVED** [review]

File: services/prometheus/prometheus.go [hunk 5]
```diff
@@ -27,8 +30,26 @@ import (
...
-func AddScrapeConfigs(l *logrus.Entry, cfg *config.Config, q *reform.Querier, s *models.MetricsResolutions) error {
-	agents, err := q.SelectAllFrom(models.AgentTable, "WHERE NOT disabled AND listen_port IS NOT NULL ORDER BY agent_type, agent_id")
+// AddScrapeConfigs - adds agents scrape configuration to given scrape config,
+// agent filter can be used for agents filtering.
+func AddScrapeConfigs(l *logrus.Entry, cfg *config.Config, q *reform.Querier, s *models.MetricsResolutions, filter models.AgentFilters) error {
+	var (
+		args       []interface{}
+		conditions []string
+	)
+	if filter.PMMAgentID != "" {
```
Comment (BupycHuk) [inline services/prometheus/prometheus.go:40]: can we move this part to models package?

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Please fix conflicts

File: services/prometheus/prometheus.go (hunk 5 above)
Comment (f41gh7) [reply services/prometheus/prometheus.go:40]: where clause generation?

Comment (BupycHuk) [reply services/prometheus/prometheus.go:40]: Yes

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/management/node.go [hunk 6]
```diff
@@ -120,8 +120,14 @@ func (s *NodeService) Register(ctx context.Context, req *managementpb.RegisterNo
 			return err
 		}
 		res.PmmAgent = a.(*inventorypb.PMMAgent)
-
-		_, err = models.CreateNodeExporter(tx.Querier, pmmAgent.AgentID, nil)
+		if _, err := models.CreateAgent(tx.Querier, models.VMAgentType, &models.CreateAgentParams{
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/management/node.go:123]: we check only later, when sending StateReques to pmm-agent.

File: services/agents/registry.go [hunk 7]
```diff
@@ -462,6 +473,17 @@ func (r *Registry) SendSetStateRequest(ctx context.Context, pmmAgentID string) {
 				return
 			}
 			rdsExporters[node] = row
+		case models.ExternalExporterType:
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/agents/registry.go:476]: below is mistaken place for ignore
https://github.com/percona/pmm-managed/blob/37300031a5daa4b65883d7070a94f8e7a14fa1f5/services/agents/registry.go#L519
there is switch only for those agents: ```case models.MySQLdExporterType, models.MongoDBExporterType, models.PostgresExporterType, models.ProxySQLExporterType,
			models.QANMySQLPerfSchemaAgentType, models.QANMySQLSlowlogAgentType, models.QANMongoDBProfilerAgentType, models.QANPostgreSQLPgStatementsAgentType,
			models.QANPostgreSQLPgStatMonitorAgentType:```

[the remark this answers is not in the document]
Comment (BupycHuk) [reply services/agents/registry.go:476]: Please recheck. It's on the same level as case you mentioned, not inside that

File: services/converters.go [hunk 8]
```diff
@@ -351,19 +358,36 @@ func ToAPIAgent(q *reform.Querier, agent *models.Agent) (inventorypb.Agent, erro
...
+			ListenPort:         uint32(pointer.GetUint16(agent.ListenPort)),
+			CustomLabels:       labels,
+			PushMetricsEnabled: agent.PushMetrics,
+		}, nil
+	case models.VMAgentType:
+		return &inventorypb.VMAgent{
+			AgentId:    agent.AgentID,
+			PmmAgentId: pointer.GetString(agent.PMMAgentID),
+			Status:     inventorypb.AgentStatus(inventorypb.AgentStatus_value[agent.Status]),
 		}, nil
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/converters.go:391]: Yep, VMAgent inventory type doesnt have PushMetricsEnabled field.

File: services/agents/registry.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/agents/registry.go:476]: hmm,  go forbids the duplicate values inside one switch case.

File: services/agents/registry.go [hunk 9]
```diff
@@ -493,10 +515,10 @@ func (r *Registry) SendSetStateRequest(ctx context.Context, pmmAgentID string) {
 				builtinAgents[row.AgentID] = qanPostgreSQLPgStatementsAgentConfig(service, row)
 			case models.QANPostgreSQLPgStatMonitorAgentType:
 				builtinAgents[row.AgentID] = qanPostgreSQLPgStatMonitorAgentConfig(service, row)
-			}
 
-		case models.ExternalExporterType:
-			// ignore
+			case models.ExternalExporterType:
```
Comment (BupycHuk) [inline services/agents/registry.go:519]: then we don't need it here

File: models/database_test.go [hunk 10]
```diff
@@ -277,44 +277,6 @@ func TestDatabaseChecks(t *testing.T) {
 			})
 		})
 
-		t.Run("runs_on_node_id_only_for_pmm_agent_and_external", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply models/database_test.go:280]: fixed

File: models/agent_helpers.go [hunk 11]
```diff
@@ -451,6 +502,9 @@ func ChangeAgent(q *reform.Querier, agentID string, params *ChangeCommonAgentPar
 			row.Disabled = false
 		}
 	}
+	if params.DisablePushMetrics != nil {
+		row.PushMetrics = !(*params.DisablePushMetrics)
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply models/agent_helpers.go:529]: added logic with externalExporter updates.

File: models/agent_helpers.go [hunk 12]
```diff
@@ -313,17 +345,31 @@ type CreateExternalExporterParams struct {
...
 	if err := checkUniqueAgentID(q, id); err != nil {
 		return nil, err
 	}
+	// with push metrics we have to detect pmm_agent_id for external exporter.
+	if params.PushMetrics {
+		agentIDs, err := FindPMMAgentsRunningOnNode(q, params.RunsOnNodeID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "cannot find pmm_agent for external exporter with push_metrics")
+		}
+		pmmAgentID = pointer.ToString(agentIDs[0].AgentID)
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply models/agent_helpers.go:393]: added len check with error

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**